File: _18_Linked_List_Cycle.java
public class _18_Linked_List_Cycle {
    public static void main(String[] args) {
        Solution sol = new Solution();

        // Caso 1: [3,2,0,-4], pos = 1  (ciclo hacia el nodo con valor 2)
        ListNode n1 = new ListNode(3);
        ListNode n2 = new ListNode(2);
        ListNode n3 = new ListNode(0);
        ListNode n4 = new ListNode(-4);
        n1.next = n2;
        n2.next = n3;
        n3.next = n4;
        n4.next = n2; // <-- ciclo aquí
        System.out.println("Caso 1: " + sol.hasCycle(n1)); // true

        // Caso 2: [1,2], pos = 0  (ciclo hacia el primer nodo)
        ListNode m1 = new ListNode(1);
        ListNode m2 = new ListNode(2);
        m1.next = m2;
        m2.next = m1; // <-- ciclo aquí
        System.out.println("Caso 2: " + sol.hasCycle(m1)); // true

        // Caso 3: [1], pos = -1  (sin ciclo)
        ListNode s1 = new ListNode(1);
        System.out.println("Caso 3: " + sol.hasCycle(s1)); // false
    }
}

class ListNode {
    int val;
    ListNode next;
    ListNode(int x) {
        val = x;
        next = null;
    }
}


//My solution
class Solution {
    public boolean hasCycle(ListNode head) {
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            head = head.next;       
            fast = fast.next.next;  

            if (head == fast) {
                return true;
            }
        }

        return false;
    }
}

File: _18_Linked_List_Cycle.ts


function hasCycle(head: ListNode3 | null): boolean {
    let slow: ListNode3 | null = head;
    let fast: ListNode3 | null = head;

    while (fast !== null && fast.next !== null) {
        slow = slow!.next;        
        fast = fast.next.next;     

        if (slow === fast) {     
            return true;
        }
    }

    return false; 
}

class ListNode3 {
    val: number;
    next: ListNode3 | null;
    constructor(val?: number, next?: ListNode3 | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

let node1 = new ListNode3(3);
let node2 = new ListNode3(2);
let node3 = new ListNode3(0);
let node4 = new ListNode3(-4);

node1.next = node2;
node2.next = node3;
node3.next = node4;
node4.next = node2; 

console.log(hasCycle(node1));

let n1 = new ListNode3(1);
let n2 = new ListNode3(2);
n1.next = n2;

console.log(hasCycle(n1)); 
